### DISCUSSION TITLE
What would it look like if controllers returned (interface{}, err) instead of void?

### DISCUSSION THREAD

**Discussion by bradcypert** [bradcypert/ghast#42] [Ideas]
Discussion area to decide if we should move away from controller funcs being a http handler func or if we should abstract that.

One of the benefits I see in this area is that we can make it a lot easier to avoid the "http: superfluous response.WriteHeader call" issue that can easily happen when you hit an error state when structuring your Go code like normal.

Update: 10/31/2021
One thing to think about is that we probably cant just return `(interface{}, err)` and instead would need to return something like `(int, interface{}, err)` so we could support multiple status codes.

**Comment by Jakeplusplus** [comment]
What do you envision getting passed back in the interface{}?

**Reply by bradcypert** [reply]
@jacobkastensmidt just to make sure were on the same page, interface{} is basically the Java.Object of Golang. I'm not suggesting we create an interface (although maybe we should), but instead allowing users to return any manner of items that we try to take care of serializing for them.

For example, returning a slice would try to marshall the slice. Returning a struct would try to marshall the struct. Returning a string would write the string to the responseWriter as is.

We may be able to make this more clever with generics, too!

**Reply by Jakeplusplus** [reply]
Yeah, I definitely like the idea of having the controller only formatting the response and not being in charge of actually writing the response.

**Reply by bradcypert** [reply]
Can you elaborate more on what you mean by "formatting" the response?

**Comment by bradcypert** [comment]
Thoughts on defining and returning a `Response` interface instead? I dont want to abuse multiple variable returns if we can avoid it.

**Comment by bradcypert** [comment]
ed45bb1 #48 included this work
